pkg/image/prune/imagepruner.go

```diff
@@ -53,12 +67,59 @@ type ImagePruner interface {
 
 // imagePruner implements ImagePruner.
 type imagePruner struct {
-	g         graph.Graph
-	algorithm pruneAlgorithm
+	g              graph.Graph
+	algorithm      pruneAlgorithm
+	registryPinger registryPinger
 }
 
 var _ ImagePruner = &imagePruner{}
 
+// registryPinger performs a health check against a registry.
+type registryPinger interface {
+	// ping performs a health check against registry.
+	ping(registry string) error
+}
+
+// defaultRegistryPinger implements registryPinger.
+type defaultRegistryPinger struct {
+}
+
+func (drp *defaultRegistryPinger) ping(registry string) error {
+	healthzCheck := func(proto, registry string) error {
+		healthzResponse, err := http.Get(fmt.Sprintf("%s://%s/healthz", proto, registry))
```

Comment:
Andy and I talked today about this - the general order should be: remove
references down to Images in the OpenShift side, then on the registry side
remove layers down to manifests, then remove manifests, then remove
images.  That's forward compatible with garbage collection.  That order is
fail safe.

On Wed, Jul 1, 2015 at 7:51 AM, David Eads <notifications@github.com> wrote:

> In pkg/image/prune/imagepruner.go
> <https://github.com/openshift/origin/pull/3515#discussion_r33670831>:
>
> >  }
> >
> >  var _ ImagePruner = &imagePruner{}
> >
> > +// registryPinger performs a health check against a registry.
> > +type registryPinger interface {
> > +	// ping performs a health check against registry.
> > +	ping(registry string) error
> > +}
> > +
> > +// defaultRegistryPinger implements registryPinger.
> > +type defaultRegistryPinger struct {
> > +}
> > +
> > +func (drp *defaultRegistryPinger) ping(registry string) error {
> > +	healthzCheck := func(proto, registry string) error {
> > +		healthzResponse, err := http.Get(fmt.Sprintf("%s://%s/healthz", proto, registry))
>
>  Hrm - I'm confused as to what we are doing here. What's the rationale
> behind a probe, vs failing the prune if the layers can't be removed?
>
> If I recall correctly, since the information about exactly what needs to
> be removed in the registry is only being built in a forward direction and
> not stored, if we're unable to contact the registry the registry pieces
> will be orphaned. Rather than orphan registry data with no hope of finding
> it, it seems better to fail early.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3515/files#r33670831>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift